classes/migration/upgrade/v3_4_0/PreflightCheckMigration.php

```diff
@@ -31,745 +34,1137 @@ abstract protected function getContextTable(): string;
-            foreach ($orphanedIds as $announcementId) {
-                $this->_installer->log("Removing orphaned settings for missing announcement ID {$announcementId}");
-                DB::table('announcement_settings')->where('announcement_id', '=', $announcementId)->delete();
-            }
+        $this->addTableProcessor('genres', function (): int {
+            $affectedRows = 0;
+            // Depends directly on ~1 entities: context_id->context_table.context_id
+            // Custom field (not found in at least one of the softwares)
+            $affectedRows += $this->deleteRequiredReference('genres', 'context_id', $this->getContextTable(), $this->getContextKeyField());
+            return $affectedRows;
+        });
 
-            // Clean orphaned category data
-            $orphanedIds = DB::table('categories AS ca')->leftJoin($this->getContextTable() . ' AS c', 'ca.context_id', '=', 'c.' . $this->getContextKeyField())->whereNull('c.' . $this->getContextKeyField())->distinct()->pluck('ca.category_id');
-            foreach ($orphanedIds as $categoryId) {
-                $this->_installer->log("Removing orphaned category ID {$categoryId} with no matching context ID.");
-                DB::table('categories')->where('category_id', '=', $categoryId)->delete();
-            }
-            // Clean orphaned category_setting entries
-            $orphanedIds = DB::table('category_settings AS cs')->leftJoin('categories AS c', 'cs.category_id', '=', 'c.category_id')->whereNull('c.category_id')->distinct()->pluck('cs.category_id');
-            foreach ($orphanedIds as $categoryId) {
-                $this->_installer->log("Removing orphaned settings for missing category ID {$categoryId}");
-                DB::table('category_settings')->where('category_id', '=', $categoryId)->delete();
-            }
+        $this->addTableProcessor('navigation_menu_item_assignments', function (): int {
+            $affectedRows = 0;
+            // Depends directly on ~3 entities: navigation_menu_id->navigation_menus.navigation_menu_id navigation_menu_item_id->navigation_menu_items.navigation_menu_item_id parent_id->navigation_menu_item_assignments.navigation_menu_item_assignment_id
+            $affectedRows += $this->deleteRequiredReference('navigation_menu_item_assignments', 'navigation_menu_item_id', 'navigation_menu_items', 'navigation_menu_item_id');
+            $affectedRows += $this->deleteRequiredReference('navigation_menu_item_assignments', 'navigation_menu_id', 'navigation_menus', 'navigation_menu_id');
+            $affectedRows += $this->deleteOptionalReference('navigation_menu_item_assignments', 'parent_id', 'navigation_menu_item_assignments', 'navigation_menu_item_assignment_id', $this->ignoreZero('parent_id'));
```

Comment:
I think this is the wrong foreign key reference -- `parent_id` refers to an entry in `navigation_menu_items`, not `navigation_menu_item_assignments`. Also, this FK constraint is slated for later (https://github.com/pkp/pkp-lib/issues/8333) so cleanup shouldn't be done now -- it will just need to be done again, before the FK constraint is added.

Reply:
Yep, the cleanup will have to be redone.
I decided to preserve the cleanups for the future foreign keys, but I can drop all of them.

Reply:
OK, but the more critical issue is that this is checking the wrong table, and deleting entries that are not invalid.

Reply:
Definitely yes, the automation is "right", the problem came from here (blindly thought that `parent_id` referred to itself):
https://github.com/pkp/pkp-lib/pull/9076/commits/a8154863bec82d7e06e9980bbdacb85546a186d0#diff-8723a186af1085addb70eb012702a54559d8b36a3a2def7900edbf3f80512e96R80

Reply:
These are the "extra" cleanups:
user_groups.context_id
filters.parent_filter_id
filters.context_id
navigation_menu_item_assignments.parent_id
navigation_menu_items.context_id
email_log.sender_id
navigation_menus.context_id
plugin_settings.context_id